codec-redis/src/main/java/io/netty/handler/codec/redis/ArrayRedisMessage.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.redis;
+
+import io.netty.util.AbstractReferenceCounted;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.StringUtil;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Arrays of <a href="http://redis.io/topics/protocol">RESP</a>.
+ */
+public class ArrayRedisMessage extends AbstractReferenceCounted implements RedisMessage {
+
+    private final List<RedisMessage> children;
+
+    private ArrayRedisMessage() {
+        children = Collections.emptyList();
+    }
+
+    /**
+     * Creates a {@link ArrayRedisMessage} for the given {@code content}.
+     *
+     * @param children the children.
+     */
+    public ArrayRedisMessage(List<RedisMessage> children) {
+        // do not retain here. children are already retained when created.
+        this.children = ObjectUtil.checkNotNull(children, "children");
+    }
+
+    /**
+     * Get children of this Arrays. It can be null or empty.
+     *
+     * @return list of {@link RedisMessage}s.
+     */
+    public List<RedisMessage> children() {
+        return children;
+    }
+
+    /**
+     * Returns whether the content of this message is {@code null}.
+     *
+     * @return indicates whether the content of this message is {@code null}.
+     */
+    public boolean isNull() {
```

Comment:
@Scottmitch , I summarized my opinions.

### Client's request-side

As I know, in client's request-side, there are no send null array cases. Anyway, when I tried this, error messages are so different. Redis server distinguish type for requests.

```
// ["set", "k1", "test1"]
$ python -c 'print("*3\r\n$3\r\nset\r\n$2\r\nk1\r\n$5\r\ntest1\r\n")' | nc 127.0.0.1 6379
+OK

// ["set", "k1", null(bulk string)] (Actually, We cannot store null. Instead delete key.)
$ python -c 'print("*3\r\n$3\r\nset\r\n$2\r\nk1\r\n$-1\r\n")' | nc 127.0.0.1 6379
-ERR Protocol error: invalid bulk length

// ["set", "k1", null(array)]
$ python -c 'print("*3\r\n$3\r\nset\r\n$2\r\nk1\r\n*-1\r\n")' | nc 127.0.0.1 6379
-ERR Protocol error: expected '$', got '*'
```

### Client's response-side.
And, client's response-side, maybe we can treat null bulk string and null array to one null value. But, I don't agree for merging two nulls. Because, I think keeping symmetry is better.

### Possiblity to make Redis proxy or slave-like relay server.
If we keep symmetry, someone can make a Redis relay server using `slaveof` command and Netty with `codec-redis`. The `slaveof` command is redirection input commands to slave servers(or custom one using RESP). Or, another one can makes Redis proxy on front of Redis like a `twemproxy`.